### DISCUSSION TITLE
Import http file

### DISCUSSION THREAD

**Discussion by Huachao** [Huachao/vscode-restclient#731] [Ideas]
Copied from @rdumont https://github.com/Huachao/vscode-restclient/issues/182#issuecomment-736448476

The statement syntax
I would prefer to follow the pattern of prefixing keywords with @ inside a comment like it is done with @name. So what do you think about the following?

# @import './login.http'

GET {{url}}/api/list HTTP/1.1
Authorization: Bearer {{login.response.body.$.data.id}}
The keyword name
I see that we have a few options: include, import, require.

My personal preference is for include, since this is basically going to be treated like a preprocessor include, as if we had copied and pasted the code into the new file. Any objections?

The tasks
 Introduce include/import/require resolution to the send request command
 Update http language definition to support the new include/import/require statement
 Document usage in the readme
 (optional) offer auto-complete for included file paths
Corner cases
Nested includes: can we agree that file paths will always be relative to each file that called the @include statement?
Anything else?

**Comment by rdumont** [comment]
Thanks for bringing this to a discussion, @Huachao! I'm bringing my comment back here, but with the original formatting:

---

### The statement syntax

I would prefer to follow the pattern of prefixing keywords with `@` inside a comment, like it is done with `@name`. So what do you think about the following?

```http
# @import './login.http'

GET {{url}}/api/list HTTP/1.1
Authorization: Bearer {{login.response.body.$.data.id}}
```


### The keyword name

I see that we have a few options: `include`, `import`, `require`.

My personal preference is for `include`, since this is basically going to be treated like a preprocessor include, as if we had copied and pasted the code into the new file. Any objections?


### The tasks

- [ ] Introduce include/import/require resolution to the send request command
- [ ] Update `http` language definition to support the new include/import/require statement
- [ ] Document usage in the readme
- [ ] (optional) offer auto-complete for included file paths


### Corner cases

* Nested includes: can we agree that file paths will always be relative to each file that called the `@include` statement?

Anything else?

**Comment by Marvin-Brouwer** [comment]
I really like the idea of not having to duplicate my auth for all the files but there's one thing I'm not happy with when reading the draft.
I see the idea is to implicitly call the imported file and in that case, I don't think that `requre`, `import` or `include` describes what it's doing.

I'd expect it to actually import a callable thing.

So for example:
```text
# login.http

###
# @name GetLoginToken
###

GET https://api.someapi.com/api/v1/authenticate HTTP/1.1
Accept: application/json
Accept-Encoding: gzip
Accept-Encoding: deflate
Cache-Control: no-store, must-revalidate, proxy-revalidate, no-cache, max-age=0

{
   token: "asdfkjaoi"
}

###
```

And then import it like this:

```text
# example.http
# @import './login.http'

### 
# @name DoSomething
###

GET {{url}}/api/list HTTP/1.1
Authorization: Bearer {{GetLoginToken().response.body.$.data.id}}
Accept-Encoding: gzip
Accept-Encoding: deflate
Cache-Control: no-store, must-revalidate, proxy-revalidate, no-cache, max-age=0

###
```

The difference here is that it looks like `GetLoginToken()` which is the named request but the `()` indicates an invocation.
Which in turn also means you can decide whether to assign to a variable once or recall the auth call every request, which is very useful for refresh tokens for example,

**Comment by MominRaza** [comment]
Related update: https://github.com/Huachao/vscode-restclient/pull/1422

This discussion is addressed by the new `@set` directive, which lets you capture response values (`response.headers.*` / `response.body.*`) into shared runtime variables and reuse them across `.http` files.
